Multi-path symex checker [blocks: 3796]

peterschrammel opened the pull request:
Based on #3794, only review after the first 7 commits.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Move counterexample beautification to goto_checker

This is language-independent and can be used by
incremental goto checkers that provide traces.

peterschrammel committed:
Use goto verifier for all-properties in CBMC

peterschrammel committed:
Use goto verifier for all-properties in JBMC

peterschrammel committed:
Add multi path symex checker

multi_path_symex_only_checkert is a
bounded model checking algorithm that determines the
status of properties through symbolic execution,
constant propagation and SAT solving.

peterschrammel committed:
Clang-format

peterschrammel committed:
Output SUCCESS and FAILURE for property status

This is currently used in all the regression tests.

peterschrammel committed:
Add all properties verifier with trace storage

All-properties currently doesn't output the traces continuously,
but waits until the end. Hence, we have to store the traces.

peterschrammel committed:
Factor out output_properties_plain text

To allow reuse for output with/without traces

peterschrammel committed:
Support outputting properties into JSON streams

peterschrammel committed:
Output property results with traces

peterschrammel committed:
Fix test matching missing property description

All properties have a description now.

peterschrammel committed:
Output number of incremental goto checker iterations

in message "x out of y failed (z iterations)".

peterschrammel committed:
Do not output results if there are no properties

peterschrammel committed:
Missing line number for unwinding assertion

peterschrammel committed:
Add goto trace storage

In all properties mode, traces are currently output
at the very end. Hence, we have to store them.

peterschrammel committed:
Multi-path checker for Java requires preprocessing

peterschrammel committed:
Fix whitespace

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eb01faecc2b6cead3d160664f04e18e9499031b0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97973551

tautschnig commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -225,3 +225,53 @@ void slice(
+    irep_idt property_id = step.get_property_id();
+
+    if(property_id.empty())
+      continue;
```
This should now be impossible.

tautschnig commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -225,3 +225,53 @@ void slice(
+    if(property_id.empty())
+      continue;
+
+    // Don't set false properties; we wouldn't have traces for them.
```
What is a "false" property?

peterschrammel replied:
Rephrased.

tautschnig commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -225,3 +225,53 @@ void slice(
+    }
+    else
+    {
+      property_infot &property_info = properties.at(property_id);
```
I'd prefer something like `auto entry_pair = properties.emplace(...); if(entry_pair.second) { updated_properties.push_back(property_id); } else { property_infot &property_info = *entry_pair.first; ... }`

peterschrammel replied:
Fixed in previous PR.

tautschnig commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -225,3 +225,53 @@ void slice(
+    {
+      // This could be a NOT_CHECKED, NOT_REACHABLE or PASS,
+      // but the equation doesn't give us precise information.
+      property_pair.second.status = property_statust::PASS;
```
I don't understand what is happening here. The comment describing the function says that each status is set to `PASS` or `FAIL` for conditions that are constant, but I don't see where we check that it is constant.

tautschnig commented on src/goto-checker/multi_path_symex_only_checker.cpp:
```diff
@@ -0,0 +1,98 @@
+
+#include <util/invariant.h>
+
+#include <goto-symex/memory_model_pso.h>
```
Why this memory model?

peterschrammel replied:
Fixed.

tautschnig commented on src/goto-checker/multi_path_symex_only_checker.h:
```diff
@@ -0,0 +1,41 @@
+    ui_message_handlert &ui_message_handler,
+    abstract_goto_modelt &goto_model);
+
+  resultt operator()(propertiest &) override;
```
Documentation is required. An undocumented class with an unnamed method isn't usable.

peterschrammel replied:
Added more documentation for undocumented methods.

I won't document the operator - doxygen automatically inserts the doc from the parent class.

tautschnig commented on src/goto-checker/all_properties_verifier.h:
```diff
@@ -43,19 +43,21 @@ class all_properties_verifiert : public goto_verifiert
 protected:
   abstract_goto_modelt &goto_model;
   incremental_goto_checkerT incremental_goto_checker;
+  unsigned iterations = 1;
```
`std::size_t`

tautschnig commented on src/goto-checker/report_util.cpp:
```diff
@@ -224,13 +224,25 @@ output_properties_plain(const propertiest &properties, messaget &log)
 
 void output_properties(
   const propertiest &properties,
+  unsigned iterations,
```
`std::size_t`

tautschnig commented on src/goto-checker/report_util.cpp:
```diff
@@ -268,6 +281,7 @@ void output_properties_with_traces(
   const propertiest &properties,
   const goto_trace_storaget &traces,
   const trace_optionst &trace_options,
+  unsigned iterations,
```
`std::size_t`

tautschnig commented on src/goto-checker/report_util.h:
```diff
@@ -26,12 +26,14 @@ void report_error(ui_message_handlert &);
 
 void output_properties(
   const propertiest &properties,
+  unsigned iterations,
```
`std::size_t`

tautschnig commented on src/goto-checker/report_util.h:
```diff
@@ -26,12 +26,14 @@ void report_error(ui_message_handlert &);
   const propertiest &properties,
   const goto_trace_storaget &traces,
   const trace_optionst &trace_options,
+  unsigned iterations,
```
`std::size_t`

tautschnig reviewed (approved):
Looks ok to me except for the comments posted earlier.

peterschrammel committed:
Property ID should always be non-empty

peterschrammel committed:
Clarify comment on updated properties

peterschrammel committed:
Fix documentation

peterschrammel committed:
Fix wrong include

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0f401f5e3d418767486df7e74b54aa0d878823ee).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98244354

forejtv commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,238 @@
+    freeze_goal_variables();
+    symex_has_run = true;
+
+    log.status() << "Running " << prop_conv.decision_procedure_text()
```
Should this `log.status` come after the block end?

peterschrammel replied:
No, this is to imitate the current behavior (if I have done it correctly).

forejtv commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,238 @@
+                 << messaget::eom;
+
+    convert_symex_target_equation(equation, prop_conv, ui_message_handler);
+    update_properties_goals_from_symex_target_equation(properties);
```
Would it be possible to document what happens when this `operator()` is executed repeatedly, and whether there are any assertions that must be met on repeated execution?

The reason I am asking is that this line seems to do something to the equation depending on that argument is given as `properties`, but the line will not be executed on the second call to `operator()`, which can pass a different argument than the first call. Is this intended?

peterschrammel replied:
To @tautschnig's delight, I have now added documentation to this function to clarify this behavior.

forejtv replied:
I'm sorry, but I still don't get it. What behaviour do you expect if somebody calls `operator()` twice in a row, each time with a different argument. Say I have properties `p1` and `p2`, each individually stored in `propertiest` object `pp1` and pp2` respectively, i.e. the containers are disjoint. Then I call
```
multi_path_symex_checkert(pp1); //here I get what is happening, p1 will be set up correctly, but p2 will be untouched as far as I can tell
multi_path_symex_checkert(pp2); //here I think all setup for p2 will be skipped, and I don't see why that is ok. 
```

peterschrammel committed:
Return set of updated properties

instead of vector to make sure elements are unique

peterschrammel committed:
Split updating properties from equation

This previously performed two actions that should be done separately:
- updating properties whose status has been explicitly determined
  in the equation
- setting a status for properties that do not occur in the equation

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: e78fc517e972d34b8933cc5c104fb798d1e7d8e2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98626921
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

forejtv reviewed (commented):
Thanks for the additional documentation. See one more response to one of the comments.

Besides, would it also be able to document what `multi_path_symex_only_checkert::
operator()(propertiest &properties)	operator()` is for? I previously thought that it would be called from some objects that use SAT, but as far as I can see in the code now, it's not the case.

allredj commented on src/goto-checker/properties.cpp:
```diff
@@ -103,16 +104,33 @@ xmlt xml(const irep_idt &property_id, const property_infot &property_info)
 
-json_objectt
-json(const irep_idt &property_id, const property_infot &property_info)
+template <class json_objectT>
```
I don't yet see why templating is necessary. Hopefully I'll understand later.

peterschrammel replied:
`json_objectt` and `json_stream_object` are interface-compatible, but don't have a common parent class.

allredj commented on src/goto-checker/all_properties_verifier.h:
```diff
@@ -43,19 +43,21 @@ class all_properties_verifiert : public goto_verifiert
 protected:
   abstract_goto_modelt &goto_model;
   incremental_goto_checkerT incremental_goto_checker;
+  std::size_t iterations = 1;
```
:question:  Why start with 1?

peterschrammel replied:
Because there is always at least one iteration.

allredj reviewed (approved):
Looks fine to me, as long as @forejtv is happy (I don't have much authority on this part of the code).

smowton commented on jbmc/src/java_bytecode/java_multi_path_symex_checker.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "java_bmc_util.h"
+
+class java_multi_path_symex_checkert : public multi_path_symex_checkert
```
Can we have some comment here? What is multi-path in this context?

smowton commented on src/goto-checker/bmc_util.h:
```diff
@@ -70,12 +70,26 @@ void slice(
+///   updated properties
+/// Note: this should inspect the equation, but the equation doesn't have
+///   any useful information at the moment.
+void update_properties_status_not_checked(
```
`update_status_of_not_checked_properties`?

smowton commented on src/goto-checker/bmc_util.h:
```diff
@@ -70,12 +70,26 @@ void slice(
+  std::unordered_set<irep_idt> &updated_properties,
   const symex_target_equationt &equation);
 
+/// Sets property status for NOT_CHECKED properties.
```
What will it set them to? Under what circumstances? Is this a subset of the changes made by the function above?

smowton commented on src/goto-checker/goto_trace_storage.h:
```diff
@@ -0,0 +1,41 @@
+  std::vector<goto_tracet> traces;
+
+  // maps property ID to index in error_traces
+  std::unordered_map<irep_idt, std::size_t> property_to_trace;
```
`to_trace_index` then?

smowton commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,238 @@
+
+  auto solver_start = std::chrono::steady_clock::now();
+
+  if(!symex_has_run)
```
Name no longer accurate, since symex definitely has run now.

peterschrammel replied:
Renamed to `equation_generated`

smowton commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,238 @@
+    update_properties_goals_from_symex_target_equation(properties);
+    convert_goals();
+    freeze_goal_variables();
+    symex_has_run = true;
```
Odd place to set this

smowton commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,238 @@
+
+  prop_convt &prop_conv = solver->prop_conv();
+
+  auto solver_start = std::chrono::steady_clock::now();
```
This is really measuring conversion time, should measure symex, convert and solve times individually?

peterschrammel replied:
That's what it is doing right now - I'm replicating the current behaviour. Enhancement would need to go into a separate PR.

smowton commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  // have we got anything to check? otherwise we return DONE
+  if(!has_properties_to_check(properties))
```
I'd say actually duplicate this; sharing it between the first-time and subsequent paths actually adds complexity IMO. We could simply have something like

```
if(already_done)
  return result;

if(first_time)
{
  symex();
  if(done)
    return result;
  make_formula();
} 

solve();
```

peterschrammel replied:
Merged the blocks to make this structure more explicit.

smowton commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,238 @@
+incremental_goto_checkert::resultt multi_path_symex_checkert::
+operator()(propertiest &properties)
+{
+  resultt result;
```
Explicitly initialise? At the moment this reads like returning undefined on line 54 (and regardless it's not clear what the default result is)

peterschrammel committed:
Add goto-checker documentation

peterschrammel committed:
Do not implicitly initialize goto checker result

Makes the code more self-documenting.

peterschrammel commented:
@smowton, I've also added a `README.md` that gives a more high-level overview.

smowton commented on src/goto-checker/README.md:
```diff
@@ -0,0 +1,77 @@
+
+A goto verifier aims at determining and reporting
+the status of all or some of the properties, _independently_ of the
+actual verification algorithm (e.g. multi-path symbolic execution,
```
Maybe this is a defined term in academia, but perhaps "path merging" rather than multi-path? After all they're both multi-path, they differ in whether they treat those paths in one go or several.

smowton commented on src/goto-checker/README.md:
```diff
@@ -0,0 +1,77 @@
+  all properties and outputs results when the verification algorithm has
+  terminated. A trace ends at a violated property.
+* \ref all_propertiest: Determines the status of all properties and
+  outputs verification results continuously. In contrast to
```
incrementally?

smowton commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,248 @@
+    const auto solver_stop = std::chrono::steady_clock::now();
+    solver_runtime += std::chrono::duration<double>(solver_stop - solver_start);
+
+    log.status() << "Running " << prop_conv.decision_procedure_text()
```
Should this log message only appear first time?

peterschrammel replied:
yes

smowton commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -0,0 +1,248 @@
+
+  const auto solver_stop = std::chrono::steady_clock::now();
+  solver_runtime += std::chrono::duration<double>(solver_stop - solver_start);
+  log.status() << "Runtime decision procedure: " << solver_runtime.count()
```
While we're here: "Decision procedure runtime:"

peterschrammel replied:
https://github.com/diffblue/cbmc/issues/4013

Please drop further suggestions there, @smowton.

smowton reviewed (approved):
Some minor suggestions, otherwise looks good, thanks for the improvements!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 37ae8b15a16ad73b6c3f776f99bb41f1c0ced601).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99316896
